lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php

```diff
@@ -181,18 +185,19 @@ protected function addAttributes( \Aimeos\MShop\Order\Item\Base\Product\Iface $o
 	/**
 	 * Returns the product items for the given product IDs limited by the map of properties
 	 *
-	 * @param string[] $productIds List of product IDs
+	 * @param string[] $productsIds List of product IDs
+	 * @param string[] $productCodes List of product codes
 	 * @return \Aimeos\Map List of items implementing \Aimeos\MShop\Product\Item\Iface with IDs as keys
 	 */
-	protected function getProductItems( array $productIds ) : \Aimeos\Map
+	protected function getProductItems( array $productIds, array $productCodes ) : \Aimeos\Map
 	{
-		$manager = \Aimeos\MShop::create( $this->getContext(), 'product' );
+		$manager = \Aimeos\MShop::create($this->getContext(), 'product');
 		$search = $manager->filter( true );
-		$expr = [
-			$search->compare( '==', 'product.id', array_unique( $productIds ) ),
-			$search->getConditions(),
-		];
-		$search->setConditions( $search->and( $expr ) );
+
+		$search->add( $search->or( [
+			$search->is( 'product.id', '==', array_unique( $productIds ) ),
+			$search->is( 'product.code', '==', array_unique( $productCodes ) ),
+		] ) );
 
 		return $manager->search( $search, ['product/property'] );
```

Comment:
We could save a lot of `$products->col( null, 'product.code' )` mappings in `addAttributes()` if we already return the merged properties already here. For that, we would need an array of product.code/product.id pairs, search for both like now and return a product.id/propertiy items map. What do you think?

Reply:
The method should probably change to reflect what it returns then.

`getProductItems` should probably become `getProductProperties`.

I guess we would also want to filter the attributes by the types configured here also, which means passing the types to method.

Do you not think passing an array of code/ids into `getProductProperties` is less explicit than the current approach, as the filter will need to extract the keys and values for the relevant columns to filter on?

Reply:
We can also pass IDs and codes separately and return a map of product.code as keys and property items as values.
Your are right, we should rename the method, `getProperties()` is shorter and would fit too

Reply:
Do we need to return the map keyed as you described?

The changes ive made locally within the now labelled `getProperties` function, are to accept the types.

I map over the types and return a new map that includes the properties.

It returns something like the following:

```php
Aimeos\Map {#1780
  #list: array:3 [
    "package-weight" => Aimeos\Map {#1778
      #list: array:1 [
        182 => "0.4"
      ]
    }
    "package-net-weight" => Aimeos\Map {#1777
      #list: array:1 [
        190 => "0.4"
      ]
    }
    "package-dangerous" => Aimeos\Map {#1772
      #list: array:2 [
        191 => "0"
        192 => "0"
      ]
    }
  ]
}
```

This is then passed to `addAttributes` replacing the existing `$products` parameter, and I iterate over this and attach to the `$orderProduct`.

The tests are still passing for me with these changes - but could be a false positive due to coverage. I can publish my changes If you like?

Reply:
Yes, that should work